File: utils/Operators/FLUX_MASS.jl
function FLUX_MASS(s::Dict{String, Any})
# FLUX_MASS - Compute the inviscid mass flux for the 2D continuity equation.
#
# Evaluates the net mass flux through all four faces (east, west, south,
# north) of each interior cell using midpoint interpolation and midpoint
# quadrature, both second-order accurate on rectangular cells.
#
# Part of: Hypersonics Stability Julia Solver - Operators Module

    ## Extract arrays to avoid repeated dictionary lookups
    rho_u  = s["var"]["rho_u"]::Matrix{Float64}
    rho_v  = s["var"]["rho_v"]::Matrix{Float64}
    lr_xn  = s["mesh"]["lr_x_normal"]::Matrix{Float64}
    lr_yn  = s["mesh"]["lr_y_normal"]::Matrix{Float64}
    lr_a   = s["mesh"]["lr_area"]::Matrix{Float64}
    bt_xn  = s["mesh"]["bt_x_normal"]::Matrix{Float64}
    bt_yn  = s["mesh"]["bt_y_normal"]::Matrix{Float64}
    bt_a   = s["mesh"]["bt_area"]::Matrix{Float64}
    vol    = s["mesh"]["volume"]::Matrix{Float64}
    flux_rho = s["flux"]["rho"]::Matrix{Float64}

    ## East and west faces
    # Midpoint interpolation (2nd order on rectangles) - fused single pass
    @views temp_rho_u_face = @. (rho_u[1:end-1, 2:end-1] + rho_u[2:end, 2:end-1]) * lr_xn / 2 +
                                 (rho_v[1:end-1, 2:end-1] + rho_v[2:end, 2:end-1]) * lr_yn / 2

    # Midpoint quadrature (2nd order on rectangles)
    @views @. flux_rho = -temp_rho_u_face[2:end, :] * lr_a[2:end, :]            # Outgoing fluxes
    @views @. flux_rho = flux_rho + temp_rho_u_face[1:end-1, :] * lr_a[1:end-1, :]  # Incoming fluxes

    ## South and north faces
    # Midpoint interpolation (2nd order on rectangles) - fused single pass
    @views temp_rho_u_face2 = @. (rho_u[2:end-1, 1:end-1] + rho_u[2:end-1, 2:end]) * bt_xn / 2 +
                                  (rho_v[2:end-1, 1:end-1] + rho_v[2:end-1, 2:end]) * bt_yn / 2

    # Midpoint quadrature (2nd order on rectangles)
    @views @. flux_rho = flux_rho - temp_rho_u_face2[:, 2:end] * bt_a[:, 2:end]            # Outgoing fluxes
    @views @. flux_rho = flux_rho + temp_rho_u_face2[:, 1:end-1] * bt_a[:, 1:end-1]  # Incoming fluxes

    ## Normalise by cell volume
    @. flux_rho = flux_rho / vol

    return s
end

File: utils/Operators/FLUX_MASS.m
function s = FLUX_MASS(s)
% FLUX_MASS - Compute the inviscid mass flux for the 2D continuity equation.
%
%   Evaluates the net mass flux through all four faces (east, west, south,
%   north) of each interior cell using midpoint interpolation and midpoint
%   quadrature, both second-order accurate on rectangular cells. The
%   resulting flux is normalised by cell volume to yield the cell-averaged
%   rate of change of density.
%
% Syntax:
%   s = FLUX_MASS(s)
%
% Inputs:
%   s - Structure containing at minimum:
%                .var.rho_u, .var.rho_v      - Momentum components (including ghosts).
%                .mesh.lr_x_normal, .mesh.lr_y_normal - Left/right face unit normals.
%                .mesh.bt_x_normal, .mesh.bt_y_normal - Bottom/top face unit normals.
%                .mesh.lr_area, .mesh.bt_area - Face areas.
%                .mesh.volume                - Cell volumes.
%
% Outputs:
%   s - Updated structure with:
%                .flux.rho - Cell-averaged mass flux (interior cells).
%
% Notes:
%   - Midpoint interpolation: face value = average of adjacent cell values.
%   - Midpoint quadrature: face flux = face value * face area.
%   - Sign convention: outgoing flux is negative, incoming flux is positive.
%
% Part of: Hypersonics Stability MATLAB Solver - Operators Module

    %% East and west faces
    % Midpoint interpolation (2nd order on rectangles)
    temp_rho_u_face = (s.var.rho_u(1:end-1,2:end-1) + s.var.rho_u(2:end,2:end-1)) .* s.mesh.lr_x_normal / 2;
    temp_rho_u_face = temp_rho_u_face + (s.var.rho_v(1:end-1,2:end-1) + s.var.rho_v(2:end,2:end-1)) .* s.mesh.lr_y_normal / 2;

    % Midpoint quadrature (2nd order on rectangles)
    s.flux.rho = -temp_rho_u_face(2:end,:) .* s.mesh.lr_area(2:end,:);            % Outgoing fluxes
    s.flux.rho = s.flux.rho + temp_rho_u_face(1:end-1,:) .* s.mesh.lr_area(1:end-1,:);  % Incoming fluxes

    %% South and north faces
    % Midpoint interpolation (2nd order on rectangles)
    temp_rho_u_face2 = (s.var.rho_u(2:end-1,1:end-1) + s.var.rho_u(2:end-1,2:end)) .* s.mesh.bt_x_normal / 2;
    temp_rho_u_face2 = temp_rho_u_face2 + (s.var.rho_v(2:end-1,1:end-1) + s.var.rho_v(2:end-1,2:end)) .* s.mesh.bt_y_normal / 2;

    % Midpoint quadrature (2nd order on rectangles)
    s.flux.rho = s.flux.rho - temp_rho_u_face2(:,2:end) .* s.mesh.bt_area(:,2:end);            % Outgoing fluxes
    s.flux.rho = s.flux.rho + temp_rho_u_face2(:,1:end-1) .* s.mesh.bt_area(:,1:end-1);  % Incoming fluxes

    %% Normalise by cell volume
    s.flux.rho = s.flux.rho ./ s.mesh.volume;
end
